File: .vendor/flamewm-c-reference/src/flamewm/ui/layout.cc
#include "layout.h"

#include <algorithm>

namespace flamewm { namespace ui {

namespace {
int clampSize(int value, const Measure& m) {
    value = std::max(m.minimum, value);
    if (m.maximum >= 0) value = std::min(m.maximum, value);
    return std::max(0, value);
}

int basis(const LayoutItem& item, bool horizontal) {
    const Measure& m = horizontal ? item.spec.width : item.spec.height;
    const int content = horizontal ? item.spec.content.w : item.spec.content.h;
    if (m.mode == Fixed) return clampSize(m.value, m);
    if (m.mode == Flex) return clampSize(m.value, m);
    return clampSize(content, m);
}

int alignedOffset(int available, int size, Alignment alignment) {
    if (alignment == AlignCenter) return (available - size) / 2;
    if (alignment == AlignEnd) return available - size;
    return 0;
}

std::vector<LayoutResult> solve(const std::vector<LayoutItem>& items, const Rect& bounds,
                                bool horizontal, const LayoutOptions& options) {
    const int origin = horizontal ? bounds.x + options.padding.left : bounds.y + options.padding.top;
    const int crossOrigin = horizontal ? bounds.y + options.padding.top : bounds.x + options.padding.left;
    const int available = std::max(0, (horizontal ? bounds.w - options.padding.left - options.padding.right
                                                    : bounds.h - options.padding.top - options.padding.bottom));
    const int cross = std::max(0, horizontal ? bounds.h - options.padding.top - options.padding.bottom
                                             : bounds.w - options.padding.left - options.padding.right);
    std::vector<int> sizes(items.size(), 0);
    int used = 0;
    int flex = 0;
    int visibleCount = 0;
    for (size_t i = 0; i < items.size(); ++i) {
        if (!items[i].spec.visible) continue;
        sizes[i] = basis(items[i], horizontal);
        used += sizes[i];
        if ((items[i].spec.width.mode == Flex && horizontal) ||
            (items[i].spec.height.mode == Flex && !horizontal))
            flex += std::max(1, horizontal ? items[i].spec.width.value : items[i].spec.height.value);
        ++visibleCount;
    }
    used += std::max(0, visibleCount - 1) * options.gap;
    int extra = available - used;
    if (extra > 0 && flex > 0) {
        int given = 0;
        for (size_t i = 0; i < items.size(); ++i) if (items[i].spec.visible) {
            const Measure& m = horizontal ? items[i].spec.width : items[i].spec.height;
            if (m.mode == Flex) {
                const int weight = std::max(1, m.value);
                const int part = (extra * weight) / flex;
                sizes[i] = clampSize(sizes[i] + part, m);
                given += part;
            }
        }
        extra -= given;
    }
    std::vector<LayoutResult> result;
    int contentExtent = 0;
    for (size_t i = 0; i < sizes.size(); ++i) {
        if (items[i].spec.visible) contentExtent += sizes[i];
    }
    contentExtent += std::max(0, visibleCount - 1) * options.gap;
    const Alignment mainAlignment = horizontal ? options.horizontal : options.vertical;
    int cursor = origin + alignedOffset(available, contentExtent, mainAlignment);
    for (size_t i = 0; i < items.size(); ++i) {
        const LayoutItem& item = items[i];
        if (!item.spec.visible) { result.push_back(LayoutResult(item.id, Rect(), false)); continue; }
        const Measure& crossMeasure = horizontal ? item.spec.height : item.spec.width;
        Alignment optionAlignment = horizontal ? options.vertical : options.horizontal;
        Alignment itemAlignment = horizontal ? item.spec.vertical : item.spec.horizontal;
        const Alignment crossAlignment = optionAlignment == AlignStretch ? itemAlignment : optionAlignment;
        int crossSize = crossMeasure.mode == Flex ||
                        crossAlignment == AlignStretch
                        ? cross : basis(item, !horizontal);
        crossSize = clampSize(crossSize, crossMeasure);
        Alignment align = crossAlignment;
        if (align == AlignStretch) crossSize = clampSize(cross, crossMeasure);
        const int offset = alignedOffset(cross, crossSize, align);
        Rect r = horizontal ? Rect(cursor, crossOrigin + offset, sizes[i], crossSize)
                            : Rect(crossOrigin + offset, cursor, crossSize, sizes[i]);
        result.push_back(LayoutResult(item.id, r, true));
        cursor += sizes[i] + options.gap;
    }
    return result;
}
}

std::vector<LayoutResult> solveStack(const Stack& stack, const Rect& bounds, const LayoutOptions& options) {
    std::vector<LayoutItem> items;
    for (size_t i = 0; i < stack.count(); ++i) items.push_back(stack.at(i));
    return solve(items, bounds, stack.direction() == Stack::Horizontal, options);
}

std::vector<LayoutResult> solveGrid(const Grid& grid, const Rect& bounds, int columns, const LayoutOptions& options) {
    std::vector<LayoutResult> out;
    if (columns <= 0) return out;
    const int innerW = std::max(0, bounds.w - options.padding.left - options.padding.right);
    const int innerH = std::max(0, bounds.h - options.padding.top - options.padding.bottom);
    const int cellW = std::max(0, (innerW - (columns - 1) * options.gap) / columns);
    std::vector<int> itemRows(grid.count(), 0);
    std::vector<int> itemCols(grid.count(), 0);
    std::vector<int> rowUse(1, 0);
    for (size_t i = 0; i < grid.count(); ++i) {
        if (!grid.at(i).spec.visible) continue;
        const int span = std::max(1, std::min(grid.at(i).span, columns));
        size_t row = rowUse.size() - 1;
        while (rowUse[row] + span > columns) {
            ++row;
            if (row == rowUse.size()) rowUse.push_back(0);
        }
        itemRows[i] = static_cast<int>(row);
        itemCols[i] = rowUse[row];
        rowUse[row] += span;
    }
    const int rows = static_cast<int>(rowUse.size());
    const int cellH = rows ? std::max(0, (innerH - (rows - 1) * options.gap) / rows) : 0;
    for (size_t i = 0; i < grid.count(); ++i) {
        const LayoutItem& item = grid.at(i);
        if (!item.spec.visible) { out.push_back(LayoutResult(item.id, Rect(), false)); continue; }
        const int row = itemRows[i];
        const int col = itemCols[i];
        const int span = std::max(1, std::min(item.span, columns - col));
        const int cellX = bounds.x + options.padding.left + col * (cellW + options.gap);
        const int cellY = bounds.y + options.padding.top + row * (cellH + options.gap);
        const int cellWidth = cellW * span + options.gap * (span - 1);
        const Alignment horizontalAlignment = options.horizontal == AlignStretch
                                                  ? item.spec.horizontal : options.horizontal;
        const Alignment verticalAlignment = options.vertical == AlignStretch
                                                ? item.spec.vertical : options.vertical;
        const int w = horizontalAlignment == AlignStretch ? cellWidth
                                                            : clampSize(basis(item, true), item.spec.width);
        const int h = verticalAlignment == AlignStretch ? cellH
                                                         : clampSize(basis(item, false), item.spec.height);
        out.push_back(LayoutResult(item.id,
                                   Rect(cellX + alignedOffset(cellWidth, w, horizontalAlignment),
                                        cellY + alignedOffset(cellH, h, verticalAlignment), w, h), true));
    }
    return out;
}

std::vector<LayoutResult> solveFormRows(const FormRows& form, const Rect& bounds, int labelWidth,
                                        int rowGap, const LayoutOptions& options) {
    LayoutOptions rowOptions = options;
    rowOptions.gap = 0;
    rowOptions.padding = Insets();
    std::vector<LayoutResult> out;
    const int innerW = std::max(0, bounds.w - options.padding.left - options.padding.right);
    const int widgetW = std::max(0, innerW - std::max(0, labelWidth) - options.gap);
    const int innerH = std::max(0, bounds.h - options.padding.top - options.padding.bottom);
    const int rowH = form.rows() ? std::max(0, (innerH - static_cast<int>(form.rows() - 1) * rowGap) /
                                                   static_cast<int>(form.rows())) : 0;
    for (size_t i = 0; i < form.rows(); ++i) {
        std::vector<LayoutItem> one(1, form.itemAt(i));
        Rect row(bounds.x + options.padding.left + std::max(0, labelWidth) + options.gap,
                 bounds.y + options.padding.top + static_cast<int>(i) * (rowH + rowGap), widgetW, rowH);
        std::vector<LayoutResult> solved = solve(one, row, true, rowOptions);
        if (!solved.empty()) out.push_back(solved[0]);
    }
    return out;
}

}} // namespace flamewm::ui

File: crates/flamewm-ui-core/src/layout.rs
//! Renderer-neutral stack/grid/form layout solver ported from the native Flame UI layer.

use flamewm_api::{Rect, Size};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct Insets {
    pub left: i32,
    pub top: i32,
    pub right: i32,
    pub bottom: i32,
}

impl Insets {
    #[must_use]
    pub const fn all(value: i32) -> Self {
        Self {
            left: value,
            top: value,
            right: value,
            bottom: value,
        }
    }

    #[must_use]
    pub const fn symmetric(horizontal: i32, vertical: i32) -> Self {
        Self {
            left: horizontal,
            top: vertical,
            right: horizontal,
            bottom: vertical,
        }
    }

    #[must_use]
    pub const fn new(left: i32, top: i32, right: i32, bottom: i32) -> Self {
        Self {
            left,
            top,
            right,
            bottom,
        }
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum MeasureMode {
    Fixed,
    #[default]
    Content,
    Flex,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum Alignment {
    Start,
    Center,
    End,
    #[default]
    Stretch,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Measure {
    pub mode: MeasureMode,
    pub value: i32,
    pub minimum: i32,
    /// `None` means unbounded.
    pub maximum: Option<i32>,
}

impl Default for Measure {
    fn default() -> Self {
        Self {
            mode: MeasureMode::Content,
            value: 0,
            minimum: 0,
            maximum: None,
        }
    }
}

impl Measure {
    #[must_use]
    pub const fn fixed(value: i32) -> Self {
        Self {
            mode: MeasureMode::Fixed,
            value,
            minimum: 0,
            maximum: None,
        }
    }

    #[must_use]
    pub const fn flex(weight: i32) -> Self {
        Self {
            mode: MeasureMode::Flex,
            value: weight,
            minimum: 0,
            maximum: None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LayoutSpec {
    pub width: Measure,
    pub height: Measure,
    pub horizontal: Alignment,
    pub vertical: Alignment,
    pub visible: bool,
    pub content: Size,
}

impl Default for LayoutSpec {
    fn default() -> Self {
        Self {
            width: Measure::default(),
            height: Measure::default(),
            horizontal: Alignment::Stretch,
            vertical: Alignment::Stretch,
            visible: true,
            content: Size::default(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LayoutItem {
    pub id: String,
    pub span: i32,
    pub spec: LayoutSpec,
}

impl LayoutItem {
    #[must_use]
    pub fn content(id: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            span: 1,
            spec: LayoutSpec::default(),
        }
    }

    #[must_use]
    pub fn fill(id: impl Into<String>) -> Self {
        let mut spec = LayoutSpec::default();
        spec.width = Measure::flex(1);
        spec.height = Measure::flex(1);
        Self {
            id: id.into(),
            span: 1,
            spec,
        }
    }

    #[must_use]
    pub fn with_span(mut self, span: i32) -> Self {
        self.span = span;
        self
    }

    #[must_use]
    pub fn with_spec(mut self, spec: LayoutSpec) -> Self {
        self.spec = spec;
        self
    }

    #[must_use]
    pub fn visible(mut self, visible: bool) -> Self {
        self.spec.visible = visible;
        self
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LayoutResult {
    pub id: String,
    pub rect: Rect,
    pub visible: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LayoutOptions {
    pub gap: i32,
    pub padding: Insets,
    pub horizontal: Alignment,
    pub vertical: Alignment,
}

impl Default for LayoutOptions {
    fn default() -> Self {
        Self {
            gap: 0,
            padding: Insets::default(),
            horizontal: Alignment::Stretch,
            vertical: Alignment::Stretch,
        }
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum StackDirection {
    #[default]
    Vertical,
    Horizontal,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Stack {
    pub direction: StackDirection,
    pub items: Vec<LayoutItem>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Grid {
    pub items: Vec<LayoutItem>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct FormRows {
    pub rows: Vec<(String, LayoutItem)>,
}

#[must_use]
pub fn solve_stack(stack: &Stack, bounds: Rect, options: LayoutOptions) -> Vec<LayoutResult> {
    solve_linear(
        &stack.items,
        bounds,
        matches!(stack.direction, StackDirection::Horizontal),
        options,
    )
}

#[must_use]
pub fn solve_grid(
    grid: &Grid,
    bounds: Rect,
    columns: i32,
    options: LayoutOptions,
) -> Vec<LayoutResult> {
    if columns <= 0 {
        return Vec::new();
    }
    let inner_width = (bounds.width - options.padding.left - options.padding.right).max(0);
    let inner_height = (bounds.height - options.padding.top - options.padding.bottom).max(0);
    let cell_width = ((inner_width - (columns - 1) * options.gap) / columns).max(0);

    let mut positions = vec![None; grid.items.len()];
    let mut row_use = vec![0_i32];
    for (index, item) in grid.items.iter().enumerate() {
        if !item.spec.visible {
            continue;
        }
        let span = item.span.max(1).min(columns);
        let mut row = row_use.len() - 1;
        while row_use[row] + span > columns {
            row += 1;
            if row == row_use.len() {
                row_use.push(0);
            }
        }
        let column = row_use[row];
        row_use[row] += span;
        positions[index] = Some((i32::try_from(row).unwrap_or(i32::MAX), column));
    }
    let rows = i32::try_from(row_use.len()).unwrap_or(i32::MAX);
    let cell_height = if rows == 0 {
        0
    } else {
        ((inner_height - (rows - 1) * options.gap) / rows).max(0)
    };

    grid.items
        .iter()
        .enumerate()
        .map(|(index, item)| {
            let Some((row, column)) = positions[index] else {
                return LayoutResult {
                    id: item.id.clone(),
                    rect: Rect::default(),
                    visible: false,
                };
            };
            let span = item.span.max(1).min(columns - column);
            let cell_x = bounds.x + options.padding.left + column * (cell_width + options.gap);
            let cell_y = bounds.y + options.padding.top + row * (cell_height + options.gap);
            let cell_extent = cell_width * span + options.gap * (span - 1);
            let horizontal = if options.horizontal == Alignment::Stretch {
                item.spec.horizontal
            } else {
                options.horizontal
            };
            let vertical = if options.vertical == Alignment::Stretch {
                item.spec.vertical
            } else {
                options.vertical
            };
            let width = if horizontal == Alignment::Stretch {
                cell_extent
            } else {
                clamp_measure(basis(item, true), item.spec.width)
            };
            let height = if vertical == Alignment::Stretch {
                cell_height
            } else {
                clamp_measure(basis(item, false), item.spec.height)
            };
            LayoutResult {
                id: item.id.clone(),
                rect: Rect::new(
                    cell_x + aligned_offset(cell_extent, width, horizontal),
                    cell_y + aligned_offset(cell_height, height, vertical),
                    width,
                    height,
                ),
                visible: true,
            }
        })
        .collect()
}

#[must_use]
pub fn solve_form_rows(
    form: &FormRows,
    bounds: Rect,
    label_width: i32,
    row_gap: i32,
    options: LayoutOptions,
) -> Vec<LayoutResult> {
    if form.rows.is_empty() {
        return Vec::new();
    }
    let inner_width = (bounds.width - options.padding.left - options.padding.right).max(0);
    let widget_width = (inner_width - label_width.max(0) - options.gap).max(0);
    let inner_height = (bounds.height - options.padding.top - options.padding.bottom).max(0);
    let row_count = i32::try_from(form.rows.len()).unwrap_or(i32::MAX);
    let row_height = ((inner_height - (row_count - 1) * row_gap) / row_count).max(0);

    form.rows
        .iter()
        .enumerate()
        .filter_map(|(index, (_, item))| {
            let index = i32::try_from(index).ok()?;
            let row = Rect::new(
                bounds.x + options.padding.left + label_width.max(0) + options.gap,
                bounds.y + options.padding.top + index * (row_height + row_gap),
                widget_width,
                row_height,
            );
            let mut row_options = options;
            row_options.gap = 0;
            row_options.padding = Insets::default();
            solve_linear(std::slice::from_ref(item), row, true, row_options)
                .into_iter()
                .next()
        })
        .collect()
}

fn solve_linear(
    items: &[LayoutItem],
    bounds: Rect,
    horizontal: bool,
    options: LayoutOptions,
) -> Vec<LayoutResult> {
    let origin = if horizontal {
        bounds.x + options.padding.left
    } else {
        bounds.y + options.padding.top
    };
    let cross_origin = if horizontal {
        bounds.y + options.padding.top
    } else {
        bounds.x + options.padding.left
    };
    let available = if horizontal {
        bounds.width - options.padding.left - options.padding.right
    } else {
        bounds.height - options.padding.top - options.padding.bottom
    }
    .max(0);
    let cross = if horizontal {
        bounds.height - options.padding.top - options.padding.bottom
    } else {
        bounds.width - options.padding.left - options.padding.right
    }
    .max(0);

    let mut sizes = vec![0; items.len()];
    let mut used = 0;
    let mut flex_weight = 0;
    let mut visible_count = 0;
    for (index, item) in items.iter().enumerate() {
        if !item.spec.visible {
            continue;
        }
        sizes[index] = basis(item, horizontal);
        used += sizes[index];
        let measure = if horizontal {
            item.spec.width
        } else {
            item.spec.height
        };
        if measure.mode == MeasureMode::Flex {
            flex_weight += measure.value.max(1);
        }
        visible_count += 1;
    }
    used += (visible_count - 1).max(0) * options.gap;
    let extra = available - used;
    if extra > 0 && flex_weight > 0 {
        for (index, item) in items.iter().enumerate() {
            if !item.spec.visible {
                continue;
            }
            let measure = if horizontal {
                item.spec.width
            } else {
                item.spec.height
            };
            if measure.mode == MeasureMode::Flex {
                let part = extra * measure.value.max(1) / flex_weight;
                sizes[index] = clamp_measure(sizes[index] + part, measure);
            }
        }
    }

    let content_extent = sizes
        .iter()
        .zip(items)
        .filter(|(_, item)| item.spec.visible)
        .map(|(size, _)| *size)
        .sum::<i32>()
        + (visible_count - 1).max(0) * options.gap;
    let main_alignment = if horizontal {
        options.horizontal
    } else {
        options.vertical
    };
    let mut cursor = origin + aligned_offset(available, content_extent, main_alignment);
    let mut result = Vec::with_capacity(items.len());

    for (index, item) in items.iter().enumerate() {
        if !item.spec.visible {
            result.push(LayoutResult {
                id: item.id.clone(),
                rect: Rect::default(),
                visible: false,
            });
            continue;
        }
        let cross_measure = if horizontal {
            item.spec.height
        } else {
            item.spec.width
        };
        let option_alignment = if horizontal {
            options.vertical
        } else {
            options.horizontal
        };
        let item_alignment = if horizontal {
            item.spec.vertical
        } else {
            item.spec.horizontal
        };
        let cross_alignment = if option_alignment == Alignment::Stretch {
            item_alignment
        } else {
            option_alignment
        };
        let mut cross_size =
            if cross_measure.mode == MeasureMode::Flex || cross_alignment == Alignment::Stretch {
                cross
            } else {
                basis(item, !horizontal)
            };
        cross_size = clamp_measure(cross_size, cross_measure);
        if cross_alignment == Alignment::Stretch {
            cross_size = clamp_measure(cross, cross_measure);
        }
        let offset = aligned_offset(cross, cross_size, cross_alignment);
        let rect = if horizontal {
            Rect::new(cursor, cross_origin + offset, sizes[index], cross_size)
        } else {
            Rect::new(cross_origin + offset, cursor, cross_size, sizes[index])
        };
        result.push(LayoutResult {
            id: item.id.clone(),
            rect,
            visible: true,
        });
        cursor += sizes[index] + options.gap;
    }
    result
}

fn basis(item: &LayoutItem, horizontal: bool) -> i32 {
    let measure = if horizontal {
        item.spec.width
    } else {
        item.spec.height
    };
    let content = if horizontal {
        item.spec.content.width
    } else {
        item.spec.content.height
    };
    match measure.mode {
        MeasureMode::Fixed | MeasureMode::Flex => clamp_measure(measure.value, measure),
        MeasureMode::Content => clamp_measure(content, measure),
    }
}

fn clamp_measure(value: i32, measure: Measure) -> i32 {
    let value = value.max(measure.minimum);
    measure
        .maximum
        .map_or(value, |maximum| value.min(maximum))
        .max(0)
}

fn aligned_offset(available: i32, size: i32, alignment: Alignment) -> i32 {
    match alignment {
        Alignment::Center => (available - size) / 2,
        Alignment::End => available - size,
        Alignment::Start | Alignment::Stretch => 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn horizontal_flex_consumes_remaining_width() {
        let mut stack = Stack {
            direction: StackDirection::Horizontal,
            items: Vec::new(),
        };
        let mut fixed = LayoutItem::content("fixed");
        fixed.spec.width = Measure::fixed(40);
        fixed.spec.height = Measure::fixed(20);
        let mut flex = LayoutItem::fill("flex");
        flex.spec.height = Measure::fixed(20);
        stack.items.extend([fixed, flex]);
        let result = solve_stack(&stack, Rect::new(0, 0, 100, 20), LayoutOptions::default());
        assert_eq!(result[0].rect.width, 40);
        assert_eq!(result[1].rect.width, 60);
    }

    #[test]
    fn hidden_items_do_not_consume_gap_or_extent() {
        let stack = Stack {
            direction: StackDirection::Horizontal,
            items: vec![
                LayoutItem::fill("one"),
                LayoutItem::fill("two").visible(false),
            ],
        };
        let options = LayoutOptions {
            gap: 8,
            ..LayoutOptions::default()
        };
        let result = solve_stack(&stack, Rect::new(0, 0, 100, 20), options);
        assert_eq!(result[0].rect.width, 100);
        assert!(!result[1].visible);
    }

    #[test]
    fn grid_span_wraps_following_item_instead_of_overlapping() {
        let grid = Grid {
            items: vec![
                LayoutItem::content("a").with_span(4),
                LayoutItem::content("b"),
            ],
        };
        let result = solve_grid(&grid, Rect::new(0, 0, 100, 50), 2, LayoutOptions::default());
        assert_eq!(result[0].rect.width, 100);
        assert_eq!(result[1].rect.x, 0);
        assert!(result[1].rect.y > result[0].rect.y);
    }

    #[test]
    fn hidden_grid_items_do_not_consume_cells() {
        let grid = Grid {
            items: vec![
                LayoutItem::content("hidden").visible(false),
                LayoutItem::content("shown"),
            ],
        };
        let result = solve_grid(&grid, Rect::new(0, 0, 100, 40), 2, LayoutOptions::default());
        assert!(!result[0].visible);
        assert_eq!(result[1].rect.x, 0);
    }
}
